homeassistant/components/wallbox/calendar.py

```diff
@@ -0,0 +1,94 @@
+    CHARGER_DATA_KEY,
+    CHARGER_LAST_EVENT,
+    CHARGER_SERIAL_NUMBER_KEY,
+    DOMAIN,
+)
+from .coordinator import WallboxCoordinator, WallboxEvent
+from .entity import WallboxEntity
+
+CALENDAR_TYPE = CalendarEntityDescription(
+    key=CHARGER_CALENDAR,
+    name=None,
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the Wallbox calendar entity."""
+    coordinator = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities([WallboxCalendarEntity(coordinator, CALENDAR_TYPE)])
+
+
+class WallboxCalendarEntity(WallboxEntity, CalendarEntity):
+    """A Wallbox calendar entity."""
+
+    def __init__(
+        self, coordinator: WallboxCoordinator, description: CalendarEntityDescription
+    ) -> None:
+        """Initialize a Wallbox calendar."""
+        super().__init__(coordinator)
+        self.entity_description = description
+        self._attr_unique_id = f"{description.key}-{coordinator.data[CHARGER_DATA_KEY][CHARGER_SERIAL_NUMBER_KEY]}"
+
+    @property
+    def event(self) -> CalendarEvent | None:
+        """Return the last event."""
+        if not self.coordinator.data[CHARGER_LAST_EVENT]:
+            return None
+        return CalendarEvent(
+            summary=self.coordinator.data[CHARGER_LAST_EVENT].summary,
+            start=self.coordinator.data[CHARGER_LAST_EVENT].start,
+            end=self.coordinator.data[CHARGER_LAST_EVENT].end,
+            description=self.coordinator.data[CHARGER_LAST_EVENT].description,
+        )
+
+    # pylint: disable-next=hass-return-type
+    async def async_get_events(  # type: ignore[override]
+        self, hass: HomeAssistant, start_date: datetime, end_date: datetime
+    ) -> list[WallboxEvent]:
+        """Get all events in a specific time frame."""
+        return await self.coordinator.async_get_sessions(
+            self.coordinator.data[CHARGER_DATA_KEY][CHARGER_SERIAL_NUMBER_KEY],
+            start_date,
+            end_date,
+        )
+
+    @callback
+    def async_write_ha_state(self) -> None:
+        """Write the state to the state machine."""
+        if self.coordinator.data[CHARGER_LAST_EVENT]:
+            self._attr_extra_state_attributes = {
+                "charger_name": self.coordinator.data[CHARGER_LAST_EVENT].charger_name,
+                "username": self.coordinator.data[CHARGER_LAST_EVENT].username,
+                "session_id": self.coordinator.data[CHARGER_LAST_EVENT].session_id,
+                "currency": self.coordinator.data[CHARGER_LAST_EVENT].currency,
+                "serial_number": self.coordinator.data[
+                    CHARGER_LAST_EVENT
+                ].serial_number,
+                "energy": self.coordinator.data[CHARGER_LAST_EVENT].energy,
+                "time": self.coordinator.data[CHARGER_LAST_EVENT].time,
+                "session_cost": self.coordinator.data[CHARGER_LAST_EVENT].session_cost,
+            }
```

Comment:
What do you mean? (my knowledge of HA isn't that deep, I mostly based on other integrations, mainly Radarr)
A calendarEvent only has the following:
    start: datetime.date | datetime.datetime
    end: datetime.date | datetime.datetime
    summary: str
    description: str | None = None
    location: str | None = None

    uid: str | None = None
    recurrence_id: str | None = None
    rrule: str | None = None
    
Sure, maybe I could use uid for session ID. Should I use something else than extra_state_attributes?

Reply:
See my remark above about using WallboxEvent in the event method (instead of CalendarEvent), it contains al required data and ensures that you do not have to go back to `coordinator.data`